app/models/report_filter.rb

```diff
@@ -5,38 +5,67 @@ class ReportFilter
     resolved
     account_id
     target_account_id
-    by_target_domain
+    search_type
+    search_term
     target_origin
   ).freeze
 
+  IGNORED_PARAMS = %i(resolved search_type search_term).freeze
+
   attr_reader :params
 
   def initialize(params)
     @params = params
   end
 
   def results
-    scope = Report.unresolved
+    scope = if params[:resolved] == '-1'
+              Report.unscoped
+            elsif params[:resolved] == '1'
+              Report.resolved
+            else
+              Report.unresolved
+            end
 
     params.each do |key, value|
-      scope = scope.merge scope_for(key, value)
+      next if IGNORED_PARAMS.include? key.to_sym
+
+      new_scope = scope_for(key, value)
+      scope = scope.merge new_scope if new_scope
     end
 
+    scope = scope.merge search_scope if searching?
     scope
   end
 
+  def search_filter
+    if params[:search_term].starts_with? '@'
+      username_parts = params[:search_term].delete_prefix('@').split('@')
+      Account.where(username: username_parts[0], domain: username_parts[1])
+    else
+      Account.where(domain: params[:search_term])
+    end
+  end
+
+  def search_scope
+    case params[:search_type].to_sym
+    when :target
+      Report.where(target_account: search_filter)
+    when :source
+      Report.where(account: search_filter)
+    else
+      raise Mastodon::InvalidParameterError, "Unknown search type: #{params[:search_type]}"
+    end
+  end
+
   def scope_for(key, value)
     case key.to_sym
-    when :by_target_domain
-      Report.where(target_account: Account.where(domain: value))
```

Comment:
This is retained by a redirect in reports_controller (though we may need to handle additional query params there)